### ISSUE TITLE
meson crashes after patch for #2629

### ISSUE THREAD

**Issue by mojca** [mesonbuild/meson#2655]
I upgraded to the latest version of meson after #2629 was claimed to be fixed.

With exactly the same test cases I'm now experiencing the following error:
```
Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/bin/meson", line 37, in <module>
    sys.exit(main())
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/bin/meson", line 34, in main
    return mesonmain.run(sys.argv[1:], launcher)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mesonbuild/mesonmain.py", line 292, in run
    return mtest.run(remaining_args)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mesonbuild/mtest.py", line 625, in run
    return th.doit()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mesonbuild/mtest.py", line 365, in doit
    self.run_tests(tests)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mesonbuild/mtest.py", line 508, in run_tests
    self.drain_futures(futures)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mesonbuild/mtest.py", line 522, in drain_futures
    self.print_stats(numlen, tests, name, result.result(), i)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mesonbuild/mtest.py", line 277, in run_single_test
    (stdo, stde) = p.communicate()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/subprocess.py", line 843, in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/subprocess.py", line 1505, in _communicate
    selector.register(self.stdout, selectors.EVENT_READ)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/selectors.py", line 299, in register
    key = super().register(fileobj, events, data)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/selectors.py", line 237, in register
    key = SelectorKey(fileobj, self._fileobj_lookup(fileobj), events, data)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/selectors.py", line 224, in _fileobj_lookup
    return _fileobj_to_fd(fileobj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/selectors.py", line 39, in _fileobj_to_fd
    "{!r}".format(fileobj)) from None
ValueError: Invalid file object: <_io.BufferedReader name=10>
FAILED: meson-test 
```

I'm sorry for not being able to provide too much of additional info. If you can guide me what to check, I'll try to explore further. Last time it was clear it was about wine, this time it is no longer clear (the tests succeed when running natively on Windows in msys2). It's potentially possible that something else went wrong because I had to clean everything after upgrading to the latest version of meson.

**Comment by nirbheek** [comment]
Does this still happen?

**Comment by etscrivner** [comment]
@nirbheek Can confirm that this _was_ happening consistently with the [ksh](https://github.com/att/ast/) build in meson 0.47.0 on macOS 10.13.6 (High Sierra), but has been fixed in 0.47.1.
